### ISSUE TITLE
[cell2fate] Error with compute_and_plot_total_velocity_scvelo

### ISSUE THREAD

**Issue by jpark27** [BayraktarLab/cell2fate#18]
Dear @AlexanderAivazidis ,

Hi, Alex! Hope you been well and congrats on publication :-)

I have been playing with cell2fate and I keep having issues on following step,

` mod.compute_and_plot_total_velocity_scvelo(adata, save = results_path + data_name + 'total_velocity_plots.png', delete = False)`

![Image](https://github.com/user-attachments/assets/a1f6964c-29ca-4b58-bded-82e355cf52ce)

When I try to specify, color parameter as follows:
` mod.compute_and_plot_total_velocity_scvelo(adata, color='celltype', save = results_path + data_name + 'total_velocity_plots.png', delete = False)`

![Image](https://github.com/user-attachments/assets/c36afbf5-7104-4d80-a7de-790aec706755)

Sadly, I didn't have no luck yet with running following instead (e.g., consume a lot of memory and just killed)
`mod.compute_and_plot_total_velocity(adata, save = results_path + data_name + 'total_velocity_plots.png', delete = False)`

What shall I do to bypass this error to plot velocity?

best wishes,
Jun

**Comment by AlexanderAivazidis** [comment]
Hi Jun thanks for the issue, can you first double check that you really have 'celltype' in your adata.obs column?

**Comment by AlexanderAivazidis** [comment]
Hi Jun! thanks for the issue. You can rename the column with celltype labels in your adata object to 'clusters' or you can change the color key like this to resolve the issue:

model.compute_and_plot_total_velocity_scvelo(adata, plotting_kwargs = {"color": 'celltype'})

The whole function documentation is here:
https://cell2fate.readthedocs.io/en/latest/cell2fate.plotting.html

Best wishes,

Alexander

**Comment by jpark27** [comment]
Thanks, Alexander! It does work with specifying `plotting_kwargs` as you suggested!

If you don't mind, may I also asking how did you plot '**Figure 4d Bottom**' cell type single stacked bar plot with adata object? I couldn't locate the suitable plot function on github/tutorial?

![Image](https://github.com/user-attachments/assets/013c800b-2981-446e-9c3c-d1ef0489cedb)

bw,
Jun

**Comment by AlexanderAivazidis** [comment]
So the top plot is available by running the compare_module_activation function. The code for the bottom plot unfortunately did not make it into the package as a standard plotting function. But here is the code that I used to make that plot:

```
from typing import List, Optional
from datetime import date
import pandas as pd
import matplotlib
import matplotlib.pyplot as plt
import numpy as np
from anndata import AnnData
from pyro import clear_param_store
from scvi import REGISTRY_KEYS
from scvi.data import AnnDataManager
from scvi.data.fields import (
    CategoricalObsField,
    LayerField,
    NumericalJointObsField,
    NumericalObsField,
)
from scvi.model.base import BaseModelClass, PyroSampleMixin, PyroSviTrainMixin
from scvi.utils import setup_anndata_dsp
import pyro.distributions as dist
import torch
import scanpy as sc
import contextlib
import io
from numpy.linalg import norm
from scipy.sparse import csr_matrix
from numpy import inner
import scvelo as scv
import scipy
import gseapy as gp

from cell2fate.utils import mu_mRNA_continousAlpha_globalTime_twoStates
import cell2fate as c2f

def compare_module_activation(self, adata, chosen_modules, chosen_clusters = None, cluster_key = 'clusters', time_key = None, time_max = None, time_min = 0,
                             save = None, ncol = 1):
    n_modules = self.module.model.n_modules
    if chosen_clusters:
        subset = np.array([c in chosen_clusters for c in adata.obs[cluster_key]])
    fig, ax = plt.subplots(3, 1, figsize=(7, 5), gridspec_kw={'height_ratios': [5, 1, 1]}, sharex = True)
    for m in chosen_modules:
        T_c = torch.tensor(0.).unsqueeze(-1).unsqueeze(-1)
        Tmax = self.samples['post_sample_means']['Tmax']
        count = 0
        fraction = 0
        fraction_list = [fraction]
        T_c_list = [0]
        ss_spliced = torch.sum(torch.tensor(self.samples['post_sample_means']['A_mgON'][m,:])\
                               /torch.tensor(self.samples['post_sample_means']['gamma_g']))
        abundance = torch.sum(mu_mRNA_continousAlpha_globalTime_twoStates(
                torch.tensor(self.samples['post_sample_means']['A_mgON'][m,:]),
                torch.tensor(0.),
                torch.tensor(self.samples['post_sample_means']['beta_g']),
                torch.tensor(self.samples['post_sample_means']['gamma_g']),
                torch.tensor(self.samples['post_sample_means']['lam_mi'][m,:]),
                torch.tensor(self.samples['post_sample_means']['T_c'][:,:,0]),
                torch.tensor(np.mean(self.samples['post_sample_means']['T_mON'][:,:,m])),
                torch.tensor(np.mean(self.samples['post_sample_means']['T_mOFF'][:,:,m])),
                torch.zeros((self.module.model.n_obs, self.module.model.n_vars)))[...,1], axis = -1)
        time = self.samples['post_sample_means']['T_c'][:,:,0].flatten() - np.min(self.samples['post_sample_means']['T_c'][:,:,0].flatten())
        ax[0].scatter(time[subset], abundance[subset], s = 10, label = 'Module ' + str(m))
        ax[1].set_ylabel('Cell type')
        ax[0].set_ylabel('Total Spliced UMI Counts')
        if time_max or time_min > 0:
            ax[0].set_xlim(time_min, time_max)
        ax[0].legend(frameon=False, ncol = ncol)
        ax[0].set_title('Module Activation Across Cells In Dataset')
        ax[1].tick_params(
                axis='both',   # changes apply to the x-axis
                which='both',  # both major and minor ticks are affected
                bottom=True,   # ticks along the bottom edge are off
                top=False,     # ticks along the top edge are off
                labelbottom=True)
        # ax[1].set_xticks([])
        ax[1].set_yticks([])
        colors = np.repeat('#3ba458', len(adata.obs['clusters']))
        clusters = np.unique(adata.obs['clusters'])
        for i in range(len(clusters)):    
            colors[adata.obs['clusters'] == clusters[i]] = adata.uns['clusters_colors'][i] 
        for i,a in enumerate(time[subset]):
            ax[1].barh(y=0, width=0.1, height=0.5, left=a, color=colors[subset][i], alpha = 0.5)
        if time_key:
            time_measured = np.array(adata.obs[time_key].astype(float))
            timesteps = np.unique(adata.obs[time_key].astype(float))
            color_palette = ['red', 'yellow', 'blue', 'grey']
            colors = np.repeat('yellow', len(adata.obs['clusters']))
            for i in range(len(timesteps)):    
                colors[time_measured == timesteps[i]] = color_palette[i]
            for i,a in enumerate(time[subset]):
                ax[2].barh(y=0, width=0.1, height=0.25, left=a, color=colors[subset][i], alpha = 0.5)
        ax[1].set_ylabel('Cell type')
        ax[2].set_yticks([])
        ax[2].set_ylabel('Age \n (Days)')
        ax[2].set_xlabel('Inferred Time')
        plt.tight_layout()
        if save:
            plt.savefig(save) 
            
    if time_key:    
        from matplotlib.lines import Line2D
        custom_lines = [Line2D([0], [0], color=color_palette[i], lw=4) for i in range(len(timesteps))]
        fig, ax = plt.subplots()
        ax.legend(custom_lines, timesteps)
    if save:
        plt.savefig(save[:-4] + '_legend' + '.pdf')

compare_module_activation(mod, adata, chosen_clusters = ['Immature \n Excitatory Neuron',
                                                         'Intermediate \n Progenitor',
                                                        'Mature \n Excitatory Neuron',
                                                        'Newborn \n Excitatory Neuron',
                                                        'Radial Glia'],
                          time_key = None, chosen_modules = [0,1,2,3,4,5,6,7,8,9], time_min = 0, time_max = 120,
                         save = results_path + data_name + 'module_activation_comparison3.pdf', ncol = 2)
```

**Comment by jpark27** [comment]
It works like charm, many thanks :-)
